File: backend/services/weather.service.js
const FORECAST_URL = "https://api.open-meteo.com/v1/forecast";
const ELEVATION_URL = "https://api.open-meteo.com/v1/elevation";
const FLOOD_URL = "https://flood-api.open-meteo.com/v1/flood";

const weatherCache = new Map();
const CACHE_TTL_MS = 90 * 1000; // 90 seconds cache

async function fetchJson(url, timeoutMs = 8000) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetch(url, { signal: controller.signal });
    if (!response.ok) {
      throw new Error(`Upstream API returned ${response.status}`);
    }
    return await response.json();
  } finally {
    clearTimeout(timer);
  }
}

function sumWindow(hourly = {}, field, hoursBack, hoursAhead = 0) {
  const times = hourly.time || [];
  const values = hourly[field] || hourly.precipitation || [];
  if (!times.length) return 0;

  const now = Date.now();
  const start = now - hoursBack * 3600 * 1000;
  const end = now + hoursAhead * 3600 * 1000;
  let sum = 0;

  for (let i = 0; i < times.length; i += 1) {
    const t = Date.parse(times[i]);
    if (!Number.isFinite(t)) continue;
    if (t >= start && t <= end) {
      sum += Number(values[i] || 0);
    }
  }

  return Number(sum.toFixed(2));
}

export async function getLiveIntelligenceInputs(latitude, longitude) {
  const cacheKey = `${Number(latitude).toFixed(4)}_${Number(longitude).toFixed(4)}`;
  const cached = weatherCache.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL_MS) {
    return cached.data;
  }

  const forecastUrl = new URL(FORECAST_URL);
  forecastUrl.searchParams.set("latitude", String(latitude));
  forecastUrl.searchParams.set("longitude", String(longitude));
  forecastUrl.searchParams.set(
    "current",
    [
      "temperature_2m",
      "relative_humidity_2m",
      "surface_pressure",
      "wind_speed_10m",
      "precipitation",
      "rain",
      "soil_moisture_0_to_1cm",
      "soil_moisture_1_to_3cm",
      "soil_moisture_3_to_9cm",
      "soil_moisture_9_to_27cm"
    ].join(",")
  );
  forecastUrl.searchParams.set(
    "hourly",
    ["precipitation", "rain", "temperature_2m", "relative_humidity_2m"].join(",")
  );
  forecastUrl.searchParams.set(
    "daily",
    [
      "precipitation_sum",
      "rain_sum",
      "temperature_2m_max",
      "temperature_2m_min",
      "et0_fao_evapotranspiration"
    ].join(",")
  );
  forecastUrl.searchParams.set("past_days", "7");
  forecastUrl.searchParams.set("forecast_days", "3");
  forecastUrl.searchParams.set("timezone", "auto");

  const elevationUrl = new URL(ELEVATION_URL);
  elevationUrl.searchParams.set("latitude", String(latitude));
  elevationUrl.searchParams.set("longitude", String(longitude));

  const floodUrl = new URL(FLOOD_URL);
  floodUrl.searchParams.set("latitude", String(latitude));
  floodUrl.searchParams.set("longitude", String(longitude));
  floodUrl.searchParams.set(
    "daily",
    "river_discharge,river_discharge_mean,river_discharge_median,river_discharge_max"
  );
  floodUrl.searchParams.set("forecast_days", "3");

  const [forecastResult, elevationResult, floodResult] = await Promise.allSettled([
    fetchJson(forecastUrl),
    fetchJson(elevationUrl),
    fetchJson(floodUrl)
  ]);

  if (forecastResult.status !== "fulfilled") {
    throw new Error(
      `Weather API failed: ${forecastResult.reason?.message || "unreachable"}`
    );
  }

  const forecast = forecastResult.value || {};
  const current = forecast.current || {};
  const hourly = forecast.hourly || {};
  const daily = forecast.daily || {};

  const rainfall1h = Number(current.precipitation ?? current.rain ?? 0);
  const rainfall3h = sumWindow(hourly, "precipitation", 3);
  const rainfall6h = sumWindow(hourly, "precipitation", 6);
  const rainfall12h = sumWindow(hourly, "precipitation", 12);
  const rainfall24h = sumWindow(hourly, "precipitation", 24);
  const rainfall72h = sumWindow(hourly, "precipitation", 72);
  const forecast1h = sumWindow(hourly, "precipitation", 0, 1);
  const forecast3h = sumWindow(hourly, "precipitation", 0, 3);
  const forecast6h = sumWindow(hourly, "precipitation", 0, 6);
  const forecast12h = sumWindow(hourly, "precipitation", 0, 12);
  const forecast24h = sumWindow(hourly, "precipitation", 0, 24);
  const forecast72h = sumWindow(hourly, "precipitation", 0, 72);

  const dailySums = Array.isArray(daily.precipitation_sum)
    ? daily.precipitation_sum.map((v) => Number(v || 0))
    : [];

  // Calculate 7-day Antecedent Precipitation Index (API = sum(0.85^t * Rain_t))
  // The first 7 entries in dailySums correspond to the past 7 days
  let antecedentPrecipitationIndex = null;
  if (dailySums.length >= 7) {
    antecedentPrecipitationIndex = 0;
    for (let i = 0; i < 7; i += 1) {
      const dayRain = dailySums[6 - i] || 0; // index 6 is yesterday, index 5 is 2 days ago
      antecedentPrecipitationIndex += dayRain * Math.pow(0.85, i + 1);
    }
    antecedentPrecipitationIndex = Number(antecedentPrecipitationIndex.toFixed(2));
  } else if (rainfall24h !== null && rainfall72h !== null) {
    antecedentPrecipitationIndex = Number(((rainfall24h * 0.85) + ((rainfall72h - rainfall24h) * 0.72)).toFixed(2));
  }

  // Multi-depth soil moisture
  const sm0_1 = current.soil_moisture_0_to_1cm !== undefined ? Number(current.soil_moisture_0_to_1cm) : null;
  const sm1_3 = current.soil_moisture_1_to_3cm !== undefined ? Number(current.soil_moisture_1_to_3cm) : null;
  const sm3_9 = current.soil_moisture_3_to_9cm !== undefined ? Number(current.soil_moisture_3_to_9cm) : null;
  const sm9_27 = current.soil_moisture_9_to_27cm !== undefined ? Number(current.soil_moisture_9_to_27cm) : null;
  const rootZoneMoisture = (sm0_1 !== null && sm1_3 !== null && sm3_9 !== null && sm9_27 !== null) 
    ? Number((sm0_1 * 0.15 + sm1_3 * 0.25 + sm3_9 * 0.35 + sm9_27 * 0.25).toFixed(3)) 
    : null;

  const evapotranspiration72h = Array.isArray(daily.et0_fao_evapotranspiration)
    ? Number(daily.et0_fao_evapotranspiration.slice(-3).reduce((a, b) => a + Number(b || 0), 0).toFixed(2))
    : null;

  const elevation =
    elevationResult.status === "fulfilled" && Array.isArray(elevationResult.value?.elevation)
      ? Number(elevationResult.value.elevation[0])
      : null;

  const floodDaily =
    floodResult.status === "fulfilled" ? floodResult.value?.daily || {} : {};
  const dischargeNow = floodDaily.river_discharge?.[0] !== undefined ? Number(floodDaily.river_discharge[0]) : null;
  const dischargeMean = floodDaily.river_discharge_mean?.[0] !== undefined ? Number(floodDaily.river_discharge_mean[0]) : null;
  const dischargeMax = floodDaily.river_discharge_max?.[0] !== undefined ? Number(floodDaily.river_discharge_max[0]) : null;

  const result = {
    current: {
      rainfall: rainfall1h,
      temperature: current.temperature_2m !== undefined ? Number(current.temperature_2m) : null,
      humidity: current.relative_humidity_2m !== undefined ? Number(current.relative_humidity_2m) : null,
      pressure: current.surface_pressure !== undefined ? Number(current.surface_pressure) : null,
      wind: current.wind_speed_10m !== undefined ? Number(current.wind_speed_10m) : null,
      soilMoisture: sm0_1,
      soilMoisture0_1: sm0_1,
      soilMoisture1_3: sm1_3,
      soilMoisture3_9: sm3_9,
      soilMoisture9_27: sm9_27,
      rootZoneSoilMoisture: rootZoneMoisture
    },
    rainfall: {
      h1: rainfall1h,
      h3: rainfall3h,
      h6: rainfall6h,
      h12: rainfall12h,
      h24: rainfall24h,
      h72: rainfall72h,
      forecast1h,
      forecast3h,
      forecast6h,
      forecast12h,
      forecast24h,
      forecast72h,
      dailySums,
      antecedentPrecipitationIndex,
      evapotranspiration72h
    },
    elevation,
    flood: {
      dischargeNow,
      dischargeMean,
      dischargeMax,
      available: floodResult.status === "fulfilled"
    },
    raw: {
      forecast,
      elevation: elevationResult.status === "fulfilled" ? elevationResult.value : null,
      flood: floodResult.status === "fulfilled" ? floodResult.value : null
    }
  };

  weatherCache.set(cacheKey, { timestamp: Date.now(), data: result });
  return result;
}

export async function getWeather(latitude, longitude) {
  const live = await getLiveIntelligenceInputs(latitude, longitude);
  return {
    current: {
      rainfall: live.current.rainfall,
      temperature: live.current.temperature,
      humidity: live.current.humidity,
      pressure: live.current.pressure,
      wind: live.current.wind
    },
    forecast: {
      rainfall: live.rainfall.forecast72h,
      dailyRainfall: live.rainfall.dailySums,
      confidence: null,
      nwpSpread: null
    },
    raw: live.raw.forecast
  };
}

File: backend/app/services/weather_service.py
import asyncio
import time
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
import httpx

from app.config import settings

# In-memory weather cache (TTL: 90 seconds)
_weather_cache: Dict[str, Dict[str, Any]] = {}
CACHE_TTL_SECONDS = 90


def _sum_window(hourly: Dict[str, Any], field: str, hours_back: int, hours_ahead: int = 0) -> float:
    """Calculates rainfall summation over a sliding temporal window."""
    times = hourly.get("time", [])
    values = hourly.get(field, []) or hourly.get("precipitation", [])
    if not times or not values:
        return 0.0

    now_ms = time.time() * 1000.0
    start_ms = now_ms - hours_back * 3600.0 * 1000.0
    end_ms = now_ms + hours_ahead * 3600.0 * 1000.0

    total = 0.0
    for t_str, val in zip(times, values):
        try:
            # Parse ISO timestamp to epoch ms
            dt = datetime.fromisoformat(t_str.replace("Z", "+00:00"))
            t_ms = dt.timestamp() * 1000.0
            if start_ms <= t_ms <= end_ms and val is not None:
                total += float(val)
        except Exception:
            continue

    return round(total, 2)


async def get_live_intelligence_inputs(latitude: float, longitude: float) -> Dict[str, Any]:
    """
    Retrieves real-time physical inputs:
    - High-resolution precipitation & soil moisture from Open-Meteo Forecast
    - DEM terrain elevation from Open-Meteo Elevation
    - Modelled river discharge from GloFAS Flood API
    """
    cache_key = f"{round(latitude, 4)}_{round(longitude, 4)}"
    now_ts = time.time()

    if cache_key in _weather_cache:
        entry = _weather_cache[cache_key]
        if now_ts - entry["timestamp"] < CACHE_TTL_SECONDS:
            return entry["data"]

    forecast_params = {
        "latitude": latitude,
        "longitude": longitude,
        "current": [
            "temperature_2m",
            "relative_humidity_2m",
            "surface_pressure",
            "wind_speed_10m",
            "precipitation"
        ],
        "hourly": [
            "precipitation",
            "temperature_2m",
            "relative_humidity_2m",
            "surface_pressure",
            "wind_speed_10m",
            "soil_moisture_0_to_1cm",
            "soil_moisture_1_to_3cm",
            "soil_moisture_3_to_9cm",
            "soil_moisture_9_to_27cm",
            "et0_fao_evapotranspiration"
        ],
        "daily": [
            "precipitation_sum",
            "precipitation_hours",
            "wind_speed_10m_max"
        ],
        "timezone": "auto"
    }

    elevation_params = {
        "latitude": latitude,
        "longitude": longitude
    }

    flood_params = {
        "latitude": latitude,
        "longitude": longitude,
        "daily": [
            "river_discharge",
            "river_discharge_mean",
            "river_discharge_median",
            "river_discharge_max",
            "river_discharge_min"
        ],
        "forecast_days": 7
    }

    async with httpx.AsyncClient(timeout=8.0) as client:
        # Concurrent async requests
        forecast_task = client.get(settings.OPEN_METEO_FORECAST_URL, params=forecast_params)
        elevation_task = client.get(settings.OPEN_METEO_ELEVATION_URL, params=elevation_params)
        flood_task = client.get(settings.OPEN_METEO_FLOOD_URL, params=flood_params)

        results = await asyncio.gather(forecast_task, elevation_task, flood_task, return_exceptions=True)

    forecast_res = results[0] if not isinstance(results[0], Exception) and results[0].status_code == 200 else None
    elevation_res = results[1] if not isinstance(results[1], Exception) and results[1].status_code == 200 else None
    flood_res = results[2] if not isinstance(results[2], Exception) and results[2].status_code == 200 else None

    forecast_json = forecast_res.json() if forecast_res else {}
    elevation_json = elevation_res.json() if elevation_res else {}
    flood_json = flood_res.json() if flood_res else {}

    hourly = forecast_json.get("hourly", {})
    current = forecast_json.get("current", {})
    daily = forecast_json.get("daily", {})

    # Compute sliding rainfall windows
    r1h = _sum_window(hourly, "precipitation", 1)
    r3h = _sum_window(hourly, "precipitation", 3)
    r6h = _sum_window(hourly, "precipitation", 6)
    r12h = _sum_window(hourly, "precipitation", 12)
    r24h = _sum_window(hourly, "precipitation", 24)
    r72h = _sum_window(hourly, "precipitation", 72)
    evap72h = _sum_window(hourly, "et0_fao_evapotranspiration", 72)

    # 24h & 72h Forecast sums
    f24h = _sum_window(hourly, "precipitation", 0, 24)
    f72h = _sum_window(hourly, "precipitation", 0, 72)

    # Antecedent Precipitation Index (API 7d)
    r_day1 = _sum_window(hourly, "precipitation", 24)
    r_day2 = _sum_window(hourly, "precipitation", 48) - r_day1
    r_day3 = _sum_window(hourly, "precipitation", 72) - _sum_window(hourly, "precipitation", 48)
    api = round(r_day1 + 0.85 * max(0.0, r_day2) + 0.70 * max(0.0, r_day3), 2)

    # Soil moisture series (extract latest available reading)
    sm0_1 = hourly.get("soil_moisture_0_to_1cm", [0.28])[-1] if hourly.get("soil_moisture_0_to_1cm") else 0.28
    sm1_3 = hourly.get("soil_moisture_1_to_3cm", [0.30])[-1] if hourly.get("soil_moisture_1_to_3cm") else 0.30
    sm3_9 = hourly.get("soil_moisture_3_to_9cm", [0.32])[-1] if hourly.get("soil_moisture_3_to_9cm") else 0.32
    sm9_27 = hourly.get("soil_moisture_9_to_27cm", [0.35])[-1] if hourly.get("soil_moisture_9_to_27cm") else 0.35
    root_zone = round((float(sm0_1 or 0) * 0.1 + float(sm1_3 or 0) * 0.2 + float(sm3_9 or 0) * 0.3 + float(sm9_27 or 0) * 0.4), 3)

    # DEM Elevation
    elevation_val = elevation_json.get("elevation", [None])
    elevation_m = float(elevation_val[0]) if isinstance(elevation_val, list) and elevation_val and elevation_val[0] is not None else 65.0

    # GloFAS discharge
    flood_daily = flood_json.get("daily", {})
    discharges = [float(v) for v in flood_daily.get("river_discharge", []) if v is not None]
    means = [float(v) for v in flood_daily.get("river_discharge_mean", []) if v is not None]

    discharge_now = discharges[0] if discharges else None
    discharge_mean = means[0] if means else None

    data = {
        "rainfall": {
            "h1": r1h,
            "h3": r3h,
            "h6": r6h,
            "h12": r12h,
            "h24": r24h,
            "h72": r72h,
            "forecast1h": round(f24h / 24.0, 2),
            "forecast3h": round((f24h / 24.0) * 3, 2),
            "forecast6h": round((f24h / 24.0) * 6, 2),
            "forecast12h": round((f24h / 24.0) * 12, 2),
            "forecast24h": f24h,
            "forecast72h": f72h,
            "antecedentPrecipitationIndex": api,
            "evapotranspiration72h": evap72h
        },
        "current": {
            "temperature": current.get("temperature_2m"),
            "humidity": current.get("relative_humidity_2m"),
            "pressure": current.get("surface_pressure"),
            "wind": current.get("wind_speed_10m"),
            "soilMoisture0_1": sm0_1,
            "soilMoisture1_3": sm1_3,
            "soilMoisture3_9": sm3_9,
            "soilMoisture9_27": sm9_27,
            "rootZoneSoilMoisture": root_zone
        },
        "daily": daily,
        "elevation": elevation_m,
        "flood": {
            "dischargeNow": discharge_now,
            "dischargeMean": discharge_mean,
            "ratio": round(discharge_now / discharge_mean, 2) if (discharge_now is not None and discharge_mean and discharge_mean > 0) else None
        }
    }

    _weather_cache[cache_key] = {
        "timestamp": now_ts,
        "data": data
    }

    return data
